lib/backend/src/Obelisk/Backend.hs

```diff
@@ -263,29 +293,100 @@ renderGhcjsFrontend urlEnc ghcjsWidgets route configs f = do
+  Just wasmAssets -> scriptTag $ view _1 (wasmScripts allJsUrl wasmAssets)
 
 -- | Load the script from the given URL in a deferred script tag.
 -- This is the default method.
-deferredGhcjsScript :: Text -> FrontendWidgetT r ()
-deferredGhcjsScript allJsUrl = elAttr "script" ("type" =: "text/javascript" <> "src" =: allJsUrl <> "defer" =: "defer") blank
+deferredGhcjsScript :: GhcjsWasmAssets -> FrontendWidgetT r ()
+deferredGhcjsScript (GhcjsWasmAssets allJsUrl mWasm) = case mWasm of
+  Nothing -> elAttr "script" ("type" =: "text/javascript" <> "src" =: allJsUrl <> "defer" =: "defer") blank
+  Just wasmAssets -> scriptTag $ view _2 (wasmScripts allJsUrl wasmAssets)
+
+scriptTag :: DomBuilder t m => Text -> m ()
+scriptTag t = elAttr "script" ("type" =: "text/javascript") $ text t
+
+wasmScripts :: Text -> Text -> (Text, Text, (Int -> Text))
+wasmScripts allJsUrl wasmRoot = (preloadScript, runJsScript, delayedJsScript)
+  where
+    jsaddleJs = wasmRoot <> "/jsaddle_core.js"
+    interfaceJs = wasmRoot <> "/jsaddle_mainthread_interface.js"
+    runnerJs = wasmRoot <> "/mainthread_runner.js"
+    wasmUrl = wasmRoot <> "/frontend.wasm"
+
+    preloadScript =
+      "add_preload_tag = function (docSrc, docType) {          \
+      \  var link_tag = document.createElement('link');        \
+      \  link_tag.rel = 'preload';                             \
+      \  link_tag.as = docType;                                \
+      \  link_tag.href = docSrc;                               \
+      \  document.head.appendChild(link_tag);                  \
+      \};                                                      \
+      \if (typeof(WebAssembly) === 'undefined') {              \
+      \  add_preload_tag('" <> allJsUrl <> "', 'script');      \
+      \} else {                                                \
+      \  add_preload_tag('" <> wasmUrl <> "', 'script');       \
+      \  add_preload_tag('" <> jsaddleJs <> "', 'script');     \
+      \  add_preload_tag('" <> interfaceJs <> "', 'script');   \
+      \  add_preload_tag('" <> runnerJs <> "', 'script');      \
+      \}"
+
+    runJsScript =
+      "add_deferload_tag = function (docSrc) {       \
+      \  var tag = document.createElement('script'); \
+      \  tag.type = 'text/javascript';               \
+      \  tag.src = docSrc;                           \
+      \  tag.setAttribute = ('defer', 'defer');      \
+      \  document.body.appendChild(tag);             \
+      \};                                            \
+      \if (typeof(WebAssembly) === 'undefined') {    \
+      \  add_deferload_tag('" <> allJsUrl <> "');    \
+      \} else {                                      \
+      \  var wasmFile = '" <> wasmUrl <> "';         \
+      \  add_deferload_tag('" <> jsaddleJs <> "');   \
+      \  add_deferload_tag('" <> interfaceJs <> "'); \
+      \  add_deferload_tag('" <> runnerJs <> "');    \
```

Comment:
Why does this one only load 3 files but the preload loads 4?

Reply:
Added comment to explain this